L2 218: Create Neuron UI

lmuntanerDfinity opened the pull request:
# Motivation

Develop the UI components to create a neuron.

# Changes

* New components in `modals/CreateNeuronModal`.
* Expanded `Input` to support dark theme.
* Expanded `button` styles to support full-width.
* Expanded Identifier to hide button to copy to the clipboard.

# Tests

* New test in route Neurons to check opening the modal
* Test for `CreateNeuronModal`. The inner components are implementation details. Tested here.

mstrasinskis commented on frontend/svelte/src/lib/components/ic/Identifier.svelte:
```diff
@@ -3,20 +3,23 @@
-  </button>
+  {#if showCopyIcon}
+    <button
+      on:click={copyToClipboard}
```
`stopPropagation` in case of click event on the Card

lmuntanerDfinity replied:
> Where are the tests?
> 
> 😜

Hehehe, I first like to do a PR, to give you time to review, then I add the tests.

If you prefer to review all at once, I can always wait for the tests.

lmuntanerDfinity replied:
yes, thanks!

lmuntanerDfinity commented:
> Where are the tests?
> 
> 😜

I like to first do a PR without tests to start the review. While you review, I add the tests and also do the changes requests.

If you prefer, I can wait to have everything finished.

lmuntanerDfinity commented on frontend/svelte/src/lib/modals/CreateNeuronModal/Modal.svelte:
```diff
@@ -0,0 +1,62 @@
+  };
+
+  onMount(() => {
+    wizard?.subscribe((value) => (currentIndex = value));
```
I did it as well. I realized it adding the tests, thanks!

peterpeterparker commented:
> I like to first do a PR without tests to start the review. 

@lmuntanerDfinity create / modify the PR to "draft" then.

Will have a look tomorrow.

lmuntanerDfinity commented on frontend/svelte/src/lib/modals/CreateNeuronModal/StakeNeuron.svelte:
```diff
@@ -0,0 +1,96 @@
+      {`${formatICP(account.balance.toE8s(), { max: 4, min: 4 })}`}
+    </h4>
+    <form on:submit={createNeuron}>
+      <Input
```
the default is `number`. Which is what I use in the `amount` as well. I didn't remember it.

lmuntanerDfinity commented:
> > I like to first do a PR without tests to start the review.
> 
> @lmuntanerDfinity create / modify the PR to "draft" then.
> 
> Will have a look tomorrow.

Yes, you are right, I always forget. I will try to remember next time.

lmuntanerDfinity commented:
@peterpeterparker @nmattia @mstrasinskis ready for review!

peterpeterparker commented on frontend/svelte/src/lib/components/accounts/AccountCard.svelte:
```diff
@@ -5,12 +5,14 @@
 </script>
 
-<Card>
+<Card on:click {role}>
```
Question: Shouldn't `click` be available only if `role` is set to `button`?

TODO: add mixin `@include interaction.tappable;` when clickable

lmuntanerDfinity replied:
If you don't pass a `on:click` when using the `Card` it's not clickable. Nothing is passed to `on:click`, it would be the same as `on:click={undefined}` which is the same as not having `on:click`.

peterpeterparker replied:
Alright then, thx

peterpeterparker commented on frontend/svelte/src/lib/components/ic/Identifier.svelte:
```diff
@@ -3,20 +3,23 @@
   import { i18n } from "../../stores/i18n";
 
   export let identifier: string;
+  export let showCopyIcon: boolean = true;
```
- rename prop, it is more than an icon, it's a function. `showCopy` maybe?

- on main screen the copy is activated but anywhere else not - i.e. more often the copy action is hidden (I think, to be checked), therefore default value shall probably be `false`?

peterpeterparker commented on frontend/svelte/src/lib/components/ui/Input.svelte:
```diff
@@ -11,14 +11,20 @@
 
   export let placeholderLabelKey: string;
 
+  export let fullWidth: boolean = false;
```
shouldn't we solve this styling option with a CSS variable?

peterpeterparker replied:
Let's have a look together next week.

lmuntanerDfinity replied:
I might need this to continue with the next task, the task was to create a Neuron, I splitted in two, to make the PRs easier to review.

peterpeterparker replied:
Add a // TODO to go on and we have a look together next week.

peterpeterparker replied:
thx

peterpeterparker commented on frontend/svelte/src/lib/components/ui/Input.svelte:
```diff
@@ -11,14 +11,20 @@
 
   $: step = inputType === "number" ? step || "any" : undefined;
+
+  let dark: boolean;
```
you don't need that state I think

```
<div class="input-block" class:fullWidth class={theme}>
```

peterpeterparker replied:
> Attributes need to be unique

In HTML but svelte takes care of regrouping these.

```
<script>
	let name = 'world';
	
	const theme = 'dark';
	
	const fullWidth = true;
</script>

<h1 class={theme} class:fullWidth>Hello {name}!</h1>
```

output => `<h1 class="dark fullWidth">Hello world!</h1>`

peterpeterparker replied:
note it was just a test, maybe that's not a bad practice in svelte and such pattern is bad in svelte too?

then maybe `class={`${theme} ${fullWidth ? 'fullWidth' : ''}`}

I just fear that with time passing we've got a lot of states and `$:` so if we can spare some it might be worth it

lmuntanerDfinity replied:
yes, but not `class="input-block" class={theme}`, I changed it to `class={ `input-block ${theme} `}`

peterpeterparker replied:
Alright it is ok. In svelte [doc](https://svelte.dev/docs#class_name) they have following example:

```
<!-- Multiple class toggles can be included -->
<div class:active class:inactive={!active} class:isAdmin>...</div>
```

lmuntanerDfinity replied:
For booleans it's ok, not for strings.

peterpeterparker replied:
> For booleans it's ok, not for strings.

You tried out or found that in the svelte doc?

peterpeterparker commented on frontend/svelte/src/lib/components/ui/Wizard/Wizard.svelte:
```diff
@@ -11,6 +11,10 @@
     selectedStepStore.update((index) => (index > 0 ? index - 1 : index));
   }
 
+  export function subscribe(callback) {
```
same topic we already discussed but sinc it's a global store, why not using directly `selectedStepStore` in the `Modal`? it would made the code shorter

lmuntanerDfinity replied:
We would expose that we are using a store for the Wizard. I'd rather focus on the API than leaking implementation details.

Maybe @nmattia or @mstrasinskis can also help.

I have the feeling that if at some point we want to use different Wizard in the same page, we couldn't use the store as is. If we then removed the store, we would have to refactor all the code where we imported the store.

I agree that it feels a little bit hacky, I can't say I am proud of this part 😂

mstrasinskis replied:
TBH looking on:
```
$: {
    if (!unsubscribeWizard && wizard) {
      unsubscribeWizard = wizard.subscribe((value) => {
        currentIndex = value;
      });
    }
  }
```
I would prefer Wizard to dispatch the changes 🙈

lmuntanerDfinity replied:
@mstrasinskis @nmattia we need your input here if you don't mind.

mstrasinskis replied:
But actually `selectedStepStore` is already used here. Probably @peterpeterparker's suggestion is not so bad.

lmuntanerDfinity replied:
ok, I will import it directly.

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/Modal.svelte:
```diff
@@ -0,0 +1,84 @@
+  let wizard: Wizard | undefined;
+  let unsubscribeWizard: Unsubscriber | undefined;
+  let currentIndex: number = 0;
+  const chooseAccount = () => {
```
goodie - inline code:

```
const chooseAccount = () => wizard?.next();
```

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/Modal.svelte:
```diff
@@ -0,0 +1,84 @@
+  });
+
+  let titleKey: string = "select_source";
+  const titleMapper = {
```
`const titleMapper: Record<string, string>`?

lmuntanerDfinity replied:
ok

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/Modal.svelte:
```diff
@@ -0,0 +1,84 @@
+  on:nnsClose
+  theme="dark"
+  size="medium"
+  showBackButton={currentIndex === 1}
```
is there a chance that we add more steps? if yes then maybe `> 0` or even using a reusable function or else?

lmuntanerDfinity replied:
yes, there will be more steps, not added yet. But the back button is only valid on the second step (index 1), the other steps do not allow to go back.

peterpeterparker replied:
alright. maybe a comment then?

lmuntanerDfinity replied:
ok

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/Modal.svelte:
```diff
@@ -0,0 +1,84 @@
+
+  export let visible: boolean;
+
+  let main: Account | undefined;
```
that's actually the selected account right? the main or a sub-account (once we have implemented these) right? so maybe already rename the state in something like `selectedAccount`?

lmuntanerDfinity replied:
I can rename and add a TODO.

Btw, this means that the Neurons page has dependency on the Account page, at least for this part. It might make sense to start with that page as well.

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/Modal.svelte:
```diff
@@ -0,0 +1,84 @@
+  let wizard: Wizard | undefined;
+  let unsubscribeWizard: Unsubscriber | undefined;
+  let currentIndex: number = 0;
+  const chooseAccount = () => {
```
maybe add some comments / todos like "// TODO: apply accounts selection"

lmuntanerDfinity replied:
yep

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/StakeNeuron.svelte:
```diff
@@ -0,0 +1,96 @@
+  const transactionIcp = ICP.fromString("0.0001") as ICP;
+  let amount: number;
+
+  const createNeuron = () => {
```
add some todos

lmuntanerDfinity replied:
and remove console log 😅

peterpeterparker commented on frontend/svelte/src/lib/utils/icp.utils.ts:
```diff
@@ -1,9 +1,12 @@
-export const formatICP = (value: bigint): string =>
+export const formatICP = (
+  value: bigint,
+  { min, max }: { min?: number; max?: number } = { min: 8, max: 8 }
```
a couple of weeks ago before you started we, the all team in a 1pm chat, decided to always display 8 decimals in the nns rewrite. I guess you are setting 4 because it's 4 currently in production. probably something worth to discuss.

lmuntanerDfinity replied:
Yes, in prod it's only 4, shall I leave it at 8?

peterpeterparker replied:
Maybe leave at 8 and add a TODO?

I might mean, if the team agree to 4, that you will have to revert the code to 4 in a separate PR but it will make the choice explicit then, maybe not that bad?

lmuntanerDfinity replied:
ok, sounds good

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/StakeNeuron.svelte:
```diff
@@ -0,0 +1,96 @@
+    color: var(--background-contrast);
+    display: flex;
+    flex-direction: column;
+    gap: 1.4rem;
```
rem? shouldn't be a value related to `--padding` or a multiple of 5px or 10px?

lmuntanerDfinity replied:
ah, true, we don't use `rem` here, sorry. I usually use rem.

I will change it.

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/StakeNeuron.svelte:
```diff
@@ -0,0 +1,96 @@
+    flex-direction: column;
+    align-items: center;
+
+    width: 90%;
```
90% ? what is the 10% value? shouldn't be something like 100% - a known value, like `calc(100% - var(--padding))` or something?

lmuntanerDfinity replied:
Actually, I think it would be better to apply a padding to the parent instead, and use a 100% width here.

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/StakeNeuron.svelte:
```diff
@@ -0,0 +1,96 @@
+    width: 90%;
+
+    small {
+      margin-top: 2rem;
```
should be related  to`--padding` I think - no rem

lmuntanerDfinity replied:
yep, sorry

peterpeterparker replied:
no worries at all

peterpeterparker commented on frontend/svelte/src/lib/modals/Modal.svelte:
```diff
@@ -8,13 +8,18 @@
 
   let dark: boolean;
   $: dark = theme === "dark";
+  let medium: boolean;
```
can probably spare the state with `class={size}`

lmuntanerDfinity replied:
same, I thought we preferred this pattern, I will change it.

peterpeterparker commented on frontend/svelte/src/lib/modals/Modal.svelte:
```diff
@@ -117,6 +124,14 @@
     overflow: hidden;
+
+    &.medium {
+      width: 520px;
```
shall we - can we - use some mixin value @mstrasinskis recently defined?

lmuntanerDfinity replied:
the breakpoints?

Or I can create new variables `--width-modal-small`, `--width-modal-medium`, ...

peterpeterparker replied:
just thought maybe worth not have "hard coded" length in a component, so yeah either mixin or variables. @mstrasinskis  took care of the media queries recently, he can probably answer that's why i pinged it

lmuntanerDfinity replied:
done, discussed with @mstrasinskis and found a nice way to do it.

peterpeterparker replied:
cool 👍

peterpeterparker commented on frontend/svelte/src/lib/modals/Modal.svelte:
```diff
@@ -128,6 +143,8 @@
     display: grid;
     grid-template-columns: var(--icon-width) 1fr var(--icon-width);
 
+    z-index: 1;
```
use `--z-index`

lmuntanerDfinity replied:
👍

peterpeterparker reviewed (changes requested):
A couple of feedback and questions, thanks.

lmuntanerDfinity commented on frontend/svelte/src/routes/Neurons.svelte:
```diff
@@ -25,8 +23,14 @@
-  // TODO: TBD https://dfinity.atlassian.net/browse/L2-226
-  const stakeNeurons = () => alert("Stake Neurons");
+  let showStakeNeuronModal: boolean = false;
+  const stakeNeurons = () => {
```
ok, I usually used the short notation only if I cared about the return value.

I will change the pattern I usually use.

mstrasinskis commented on frontend/svelte/src/tests/lib/modals/CreateNeuronModal.spec.ts:
```diff
@@ -0,0 +1,86 @@
+    const input = container.querySelector('input[name="amount"]');
+    // Svelte generates code for listening to the `input` event
+    // https://github.com/testing-library/svelte-testing-library/issues/29#issuecomment-498055823
+    await fireEvent.input(input, { target: { value: 22 } });
```
Nice 😊

peterpeterparker commented:
Adding tests for the changes in the existing components that have been modified is something you will do once everything reviewed do I get it right?

lmuntanerDfinity commented:
> Adding tests for the changes in the existing components that have been modified is something you will do once everything reviewed do I get it right?

I already changed the tests for the `Identifier` regarding the copy icon. Also added the test in the `Neurons` route that opens the modal. Do you want me to add tests for the prop `size` in the `Modal` and also the props `theme` and `fullWidth` in `Input`? I didn't know we wanted to test all the css props as well.

A part from that, @peterpeterparker ready for another review

lmuntanerDfinity commented:
@mstrasinskis @nmattia ready for one more review

peterpeterparker commented:
@lmuntanerDfinity I personally would say yes, tests for any props particularly if they are changes but if the others are fine without for sure I can live with it.

a part from that, review is alright thanks for the changes. only thing which still remains open though is my question regarding "wizard global store and "hacky" (😉) subscribe". looking forward to the feedback of maks and nicolas about it.

mstrasinskis commented on frontend/svelte/src/lib/themes/variables.scss:
```diff
@@ -11,6 +11,14 @@
   --section-max-width: 800px;
   --footer-height: 20vh;
 
+  --small-width: 300px;
```
Because it's a global scope some variable name prefix (like `modal-`) would be useful.

lmuntanerDfinity replied:
ok, makes sense

mstrasinskis commented on frontend/svelte/src/lib/themes/variables.scss:
```diff
@@ -11,6 +11,14 @@
+  --small-width: 300px;
+  --medium-width: 300px;
+
+  @media (min-width: 768px) {
```
```scss
  @use "./mixins/media";
  ...
  @include media.min-width(medium) {
     ...
```
?

lmuntanerDfinity replied:
I didn't know I could add the mixin here, great!

lmuntanerDfinity commented:
@peterpeterparker I have refactored the `wizardStore` to the pattern we use when we expose it.

https://github.com/dfinity/nns-dapp/pull/423/files#diff-38e808e7afe7d3789d555425b9b06349a1139d67e3bac771c3897b89d0b057bfR3

I have also added extra tests. Also one on the Wizard that when unmounted, it resets the store.

@mstrasinskis please also review 😄

peterpeterparker commented on frontend/svelte/src/lib/components/ui/Wizard/wizardStore.ts:
```diff
@@ -1,3 +1,20 @@
 import { writable } from "svelte/store";
 
-export const selectedStepStore = writable(0);
+const initStore = () => {
```
Can you add a small comment about what the store does and what are the exposed functions (@nmattia will appreciate it, see other stores as example)?

lmuntanerDfinity replied:
done!

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/Modal.svelte:
```diff
@@ -0,0 +1,82 @@
+
+  onMount(() => {
+    unsubscribeWizard = wizardStore.subscribe((value) => {
+      currentIndex = value;
```
goodie / question - not sure can't you spare the `onMount` and `subscribe` by using `showBackButton={$currentIndex === 1}` in the template (-> `$`)

I can live with it, just in case it can indeed same some lines of code

lmuntanerDfinity replied:
Yes, true. I didn't need the `$`. The same way we subscribe to the `accountsStore`.

peterpeterparker commented on frontend/svelte/src/lib/components/ui/Wizard/wizardStore.ts:
```diff
@@ -1,3 +1,20 @@
+  };
+};
+
+export const wizardStore = initStore();
```
can you add tests`wizard.store.spec.ts` for this store?

lmuntanerDfinity replied:
Instead of testing the store on it sown, I added tests in `Wizard.spec.ts`.

https://github.com/dfinity/nns-dapp/pull/423/files#diff-ae65a891043a0a4f905b31d49b0f92984adbb79920a0434887b0fcb59abee1c3R38

I also added a comment in the store about the testing:

https://github.com/dfinity/nns-dapp/pull/423/files#diff-38e808e7afe7d3789d555425b9b06349a1139d67e3bac771c3897b89d0b057bfR9

peterpeterparker replied:
ok

peterpeterparker reviewed (changes requested):
Thanks for the improvements!!! Almost there 😉

lmuntanerDfinity commented:
@peterpeterparker ready for another review!

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/Modal.svelte:
```diff
@@ -0,0 +1,80 @@
+  );
+
+  let currentIndex: number = 0;
+  let unsubscribeWizard: Unsubscriber = wizardStore.subscribe(
```
`const`

lmuntanerDfinity replied:
yep

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/Modal.svelte:
```diff
@@ -0,0 +1,80 @@
+    unsubscribeAccounts();
+  });
+
+  let titleKey: string = "select_source";
```
`let titleKey: string = titleMapper[0]` so you can spare writing "select_source" twice

lmuntanerDfinity replied:
good point!

peterpeterparker commented on frontend/svelte/src/lib/modals/CreateNeuronModal/StakeNeuron.svelte:
```diff
@@ -0,0 +1,96 @@
+  import { formatICP } from "../../utils/icp.utils";
+
+  export let account: Account;
+  const transactionIcp = ICP.fromString("0.0001") as ICP;
```
`const transactionIcp: ICP = `

lmuntanerDfinity replied:
ok

peterpeterparker commented on frontend/svelte/src/lib/components/ic/Identifier.svelte:
```diff
@@ -3,20 +3,23 @@
-  >
-    <IconCopy />
-  </button>
+  {#if showCopy}
```
isn't a test missing for this new toggle?

lmuntanerDfinity replied:
Yes, it's there: https://github.com/dfinity/nns-dapp/pull/423/files#diff-2a1cb7a26e14da35ffc680918de7a4b86ead4fb68bea404b2687f84f72c5e51bR20

The first test does not check for the button, while the second one does.

The next test: "should copy identifier to clipboard" also uses this prop to check that it actually copies the text to the clipboard.

peterpeterparker replied:
The link does not point to the specific code for me. Anyway as far as I can tell there is no test that assert that the button is not rendered for the default showCopy=false value.

lmuntanerDfinity replied:
No, there is no test that when the prop is false there is no button.

IMHO, I find testing that the `false` don't do something, it's a little bit too much testing.

Sorry for the broken link, not sure what happened. Here are the tests in this branch: one to check that it renders the identifier and another one that the `showCopy` renders a button. There is also the one that the button copies the text to the clipboard.

```
  it("should render an identifier", () => {
    const { container } = render(Identifier, { props });

    const small = container.querySelector("small");
    expect(small).not.toBeNull();
    expect(small.textContent).toEqual(identifier);
  });

  it("should render an accessible button", () => {
    const { getByRole } = render(Identifier, {
      props: { identifier, showCopy: true },
    });

    const button = getByRole("button");

    expect(button).not.toBeNull();
    expect(button.getAttribute("aria-label")).toEqual(
      "Copy identifier to clipboard"
    );
  });
```

lmuntanerDfinity replied:
I added it

peterpeterparker reviewed (changes requested):
A few goodie (not must) and I think it needs a small additional test.

Afterwards I'll approve it.

lmuntanerDfinity commented:
> 

Thanks @peterpeterparker !

IMHO, those changes should be taken care with a linter.

peterpeterparker reviewed (approved):
Beside my comment about the test, good to me.
